clang/lib/Tooling/Refactoring/Rename/USRLocFinder.cpp

```diff
@@ -266,7 +265,7 @@ class RenameLocFinder : public RecursiveASTVisitor<RenameLocFinder> {
     return true;
   }
 
-  bool VisitDeclRefExpr(const DeclRefExpr *Expr) {
```

Comment:
Why are we removing const here (and other places)?

Reply:
So `RecursiveASTVisitor` passes everything around by non-`const` pointer, and `DynamicRecursiveASTVisitor` does the same. But because this doesn’t matter for CRTP, we have some visitors that take only non-`const` pointers, some that take `const` pointers, and some that mix both seemingly at random. For virtual functions of course, the type has to match exactly, so I had to remove `const` in a bunch of places.

Reply:
> How hard do you think it would be to introduce something like `MutableDynamicRecursiveASTVisitor` and preserve constness where possible?

I mean, it would basically just involve copy-pasting the entire DRAV interface+implementation and adding `const` everywhere; personally, I don’t really think it’s worth it because there aren’t that many visitors that were using `const` in the first place, and the entire RAV implementation also isn’t `const` to begin with...

Reply:
@AaronBallman @erichkeane

Reply:
Urgh, yeah... our const correctness with the recursive AST visitor is... bad.  One thing I think we'd be better off doing if we cared to make this change (rather than copy/paste) would be to make `RecursiveASTVistior` take a template parameter that says whether it is const correct-ish and apply it that way.  our parameters would have a lot of `add_const_if<DeclRefExpr, IsConst>` (where `add_const_if` is a new template we would need to add, and `IsConst` is the template parameter I referred to).

Reply:
> One thing I think we'd be better off doing if we cared to make this change (rather than copy/paste) would be to make `RecursiveASTVistior` take a template parameter that says whether it is const correct-ish and apply it that way.

That still doesn’t work too well for DynamicRecursiveASTVisitor—I guess we could make the template parameter a `bool` and then explicitly instantiate the only two possible instantiations in DynamicRecursiveASTVisitor.cpp...